### PR TITLE
${GDC} reading is now lockfree

### LINKED ISSUE #2960: NLog variables are not threadsafe
@snakefoot I see your recommendation for  gdc over var,

e.g.

> If needing to modify global variables at runtime, then it is recommended to use the GDC layout renderer instead. Because NLog variables are not threadsafe, and should be seen as readonly variables. Also GDC will behave correctly if LoggingConfiguration should change (or reloaded).

Is there something we could do to make the current nlog variables threadsafe? A lot of users find the variables more natural to use.

### PR DESCRIPTION
**Pull request by snakefoot**
GlobalDiagnosticsContext - Use copy-on-write and lockfree read

See also #3197 

Noticed that NLog 4.5.11 often showed threads in progress with GDC-Lookups. On NLog 4.6 one didn't see it because GDC is marked as ThreadAgnostic.

But if using GDC in a Layout with other LayoutRenderers that are not ThreadAgnostic, then the lock-punishment will be back.

Added GDC-Lookup to the file-target-layout and now they get stuck with NLog 4.6:
image

### CODE REVIEW

**Comment by snakefoot** [comment]
ConcurrentDictionary doesn't make the same "promise" of ordering as normal Dictionary. So JsonLayout.IncludeGdc is not producing identical output.

**Comment by snakefoot** [comment]
The ordering is important in the expanding of variables that reference other variables. So reverted back to only using ConcurrentDictionary for GlobalDiagnosticsContext

**Comment by 304NotModified** [comment]
Isn't the concurrent queue much slower then the lock?

**Comment by snakefoot** [comment]
> Isn't the concurrent queue much slower then the lock?

Not if you have two threads.

**Comment by 304NotModified** [comment]
Not sure it's coincidence this tests fails now :

NLog.UnitTests.Targets.ConcurrentFileTargetTests.SimpleConcurrentTest(numProcesses: 2, numLogs: 500, mode: "none|archive") [FAIL]
At C:\projects\nlog\run-tests.ps1:33 char:1
+ & ${env:xunit20}\xunit.console.x86.exe .\tests\NLog.UnitTests\bin\rel ...

From 
https://ci.appveyor.com/project/nlog/nlog/builds/23379399

File: tests/NLog.UnitTests/Layouts/JsonLayoutTests.cs [hunk 1]
```diff
@@ -506,7 +506,7 @@ public void IncludeGdcJsonProperties()
             <targets>
                 <target name='asyncDebug' type='AsyncWrapper' timeToSleepBetweenBatches='0'>
                 <target name='debug' type='Debug'  >
-                 <layout type=""JsonLayout"" IncludeGdc='true' ExcludeProperties='Excluded1,Excluded2'>
+                 <layout type=""JsonLayout"" IncludeGdc='true' IncludeAllProperties='true' ExcludeProperties='Excluded1,Excluded2'>
```
[the remark this answers is not in the document]
Comment (snakefoot) [reply tests/NLog.UnitTests/Layouts/JsonLayoutTests.cs]: Dictionary makes a very good promise to return items in the same order as inserted (Though not always). ConcurrentDictionary doesn't make this promise.

[the remark this answers is not in the document]
Comment (snakefoot) [reply tests/NLog.UnitTests/Layouts/JsonLayoutTests.cs]: I have modified the test so GDC only contains the first property, and rest of the properties comes from the LogEvent.

**Comment by snakefoot** [comment]
> Not sure it's coincidence this tests fails now :

Always been an unstable test. Concurrent processes writing to the same file test is not affected by this change.

**Comment by snakefoot** [comment]
I think the trouble and the surprises with ConcurrentDictionary will outweight the gain in a truly multithreaded environment.

Even using readerwriterlockslim will probably have some side-effects:

https://blogs.msdn.microsoft.com/pedram/2007/10/07/a-performance-comparison-of-readerwriterlockslim-with-readerwriterlock/

**Comment by snakefoot** [comment]
@304NotModified Now modified GDC to perform clone on write. This gives high concurrency for loggers but more allocation for writers (if they write a lot while others are reading).

**Comment by 304NotModified** [comment]
@snakefoot is this one safe for NLog 4.6.1? (So low impact)?

**Comment by snakefoot** [comment]
I think it is safe for NLog 4.6.1. One could also consider using the same approach for NLog Configuration Variables.

File: src/NLog/GlobalDiagnosticsContext.cs [hunk 2]
```diff
@@ -152,10 +142,40 @@ public static void Remove(string item)
...
+            }
+            return readOnly;
+        }
+
+        private static Dictionary<string, object> GetWritableDict(bool mustCreateNew, bool clone = true)
+        {
+            if (mustCreateNew)
             {
-                dict.Clear();
+                var newDict = new Dictionary<string, object>(clone ? _dict.Count + 1 : 0);
```
Comment (304NotModified) [inline src/NLog/GlobalDiagnosticsContext.cs]: Is it correct that clear has the init capacity of 0, and the field has the default capacity (16?). It that on purpose?

**Comment by 304NotModified** [comment]
please check https://github.com/snakefoot/NLog/pull/5 :)

File: src/NLog/GlobalDiagnosticsContext.cs [hunk 3]
```diff
@@ -69,7 +68,7 @@ public static void Set(string item, object value)
         {
             lock (dict)
             {
-                dict[item] = value;
+                GetWritableDict(dictReadOnly != null && !dict.ContainsKey(item))[item] = value;
```
[the remark this answers is not in the document]
Comment (snakefoot) [reply src/NLog/GlobalDiagnosticsContext.cs]: Yes one could probably make one of the Setters call the other.

File: src/NLog/GlobalDiagnosticsContext.cs (hunk 2 above)
Comment (snakefoot) [reply src/NLog/GlobalDiagnosticsContext.cs]: Don't understand your question. But Clear (clone=false) calls with capacity = 0 that means do not allocate internal arrays until insert happens.

Comment (304NotModified) [reply src/NLog/GlobalDiagnosticsContext.cs]: indeed, but the init doesn't set it to `0`

`private static Dictionary<string, object> _dict = new Dictionary<string, object>();`

Comment (snakefoot) [reply src/NLog/GlobalDiagnosticsContext.cs]: Still don't understand your queustion. The default constructor for Dictionary uses capacity = 0. See also https://referencesource.microsoft.com/#mscorlib/system/collections/generic/dictionary.cs,85

**Comment by snakefoot** [comment]
@304NotModified Have now extracted the clone-logic from GetWritableDict

File: src/NLog/GlobalDiagnosticsContext.cs [hunk 4]
```diff
@@ -158,21 +158,27 @@ public static void Clear()
             return readOnly;
         }
 
-        private static Dictionary<string, object> GetWritableDict(bool mustCreateNew, bool clone = true)
+        private static Dictionary<string, object> GetWritableDict(bool mustCreateNew, bool clearDictionary = false)
```
Comment (304NotModified) [inline src/NLog/GlobalDiagnosticsContext.cs]: This is a strange methode, with mustCreateNew and clear IMO. I tried to refactor that ;)

Comment (snakefoot) [reply src/NLog/GlobalDiagnosticsContext.cs]: It is very identical to MDC + MDLC - GetLogicalThreadDictionary

Comment (304NotModified) [reply src/NLog/GlobalDiagnosticsContext.cs]: well, clearDictionary  is only called once with `true`. Then it's strange to have that encapsulated in a method. It makes things a bit unnecessary complex.

Comment (snakefoot) [reply src/NLog/GlobalDiagnosticsContext.cs]: Yes I'm terrible at making code complex. But trying to improve, though I often forget. But I think https://github.com/snakefoot/NLog/pull/5 now is showing boiler-plate code for the different write-operations.

Comment (snakefoot) [reply src/NLog/GlobalDiagnosticsContext.cs]: Btw. the clearDictionary is an optimization-flag, so it knows it can skip the copying of the old dictionary.

Comment (304NotModified) [reply src/NLog/GlobalDiagnosticsContext.cs]: > showing boiler-plate code for the different write-operations.

true, set, remove and clear has code duplication, but that was already the case (in 1 line) and now those are 2 lines of duplication. Against it, CloneDictionary is much more straight-forward.

Comment (304NotModified) [reply src/NLog/GlobalDiagnosticsContext.cs]: Also GetWritableDict wasn't doing one thing, it was doing "get" but also "set" (but OK, that's also the case in MDC probably, but im trying to "fix the leak")

Comment (snakefoot) [reply src/NLog/GlobalDiagnosticsContext.cs]: With copy-on-write semantics then get-operation will sometimes be create/set. That is the nature of copy-on-write.

Comment (304NotModified) [reply src/NLog/GlobalDiagnosticsContext.cs]: from another view, do you think this line is easy to maintain?

`  GetWritableDict(_dictReadOnly != null && !_dict.ContainsKey(item))[item] = value;`

Comment (304NotModified) [reply src/NLog/GlobalDiagnosticsContext.cs]: > Yes I'm terrible good at making complex code.

I focus on having maintainable code, and unfortunately that's not always performant code :(  I think the best code for NLog is in-between :)

Comment (snakefoot) [reply src/NLog/GlobalDiagnosticsContext.cs]: Your points are very valid and valuable. Have now made another code-update for readability.

**Comment by snakefoot** [comment]
Happy to do it. And also very happy about the result :)

**Pull request merged**